### PR TITLE
Add hackbench and schbench for scheduler testing

### PR DESCRIPTION
**Pull request by jellllly420**
This PR adds two scheduler benchmarks to our CI test. Should go after #1687.

The settings of benchmarks aim at covering most scenarios:
- nr_cpu
    -  = 1 -> single core
    -  = nr_thread -> max loading
    - < nr_thread -> over loading
- task_size
    - small -> voluntary context switch
    - big -> involuntary context swtich

### CODE REVIEW

File: test/Makefile
Comment (tatetian) [inline test/Makefile:125]: Adding this line will fail most CI tests. I suggest adding it later when the new Docker image is built.

File: tools/docker/Dockerfile.jinja
Comment (tatetian) [inline tools/docker/Dockerfile.jinja:57]: Now that the Dockerfile is updated, we need to build the new Docker image. This can be done by bumping the version of Asterinas  to 0.10.1 with `tools/bump_version.sh`. The bumped version will trigger Docker image build CI.

**Review by tatetian — COMMENTED** [review]
Thanks for the attempt of adding a benchmark for task schedulers. I think this tool will be useful.

My concern is that adding the benchmark program itself would be meaningless, unless it is run as part of our benchmark CI. Do you have a plan to submit a followup PR to add it to the benchmark CI?

**Comment by jellllly420** [comment]
@tatetian Since there can be various settings when running the benchmark, I'm not sure which sets of settings is most representative to be added to our CI. I'll try to see how its other users (e.g., Linux) use it and update this PR later.

File: test/Makefile
Comment (tatetian) [reply test/Makefile:125]: > Does it mean I'd have to update the Dockerfile and bump the version in one PR 

Yes, that would be great.

**Comment by jellllly420** [comment]
@boterinas rerun

**Comment by tatetian** [comment]
This PR will be blocked until #1576 is merged. That PR is a rework of the benchmark framework. You will have to refactor this PR by following the new interface and conventions described by that PR. Sorry for the inconvenience.

**Comment by tatetian** [comment]
@jellllly420 #1576 has been merged. Check out the new benchmark framework and refactor your new benchmarks accordingly.

File: test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json
Comment (grief8) [reply test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json:8]: I have executed the shell locally. We didn't consider such a situation that a benchmark would run several times and generate multiple similar results. Current practice takes the last result as the final result, which wastes other results. Since schbench does not provide an interface for averaging, I think we should do this ourselves. We can process these data in the virtual machine, that is, `run.sh`, or in `bench_linux_and_aster.sh`. I tend to `run.sh`.

```sh
#!/bin/sh

# SPDX-License-Identifier: MPL-2.0

/benchmark/bin/schbench -F 256 -n 5 -r 90 | tee schbench_output.txt

# Extract percentile values marked with * and calculate their averages
awk '
BEGIN {
    # Initialize variables
    sum_wakeup = 0
    count_wakeup = 0
    sum_request = 0
    count_request = 0
    sum_rps = 0
    count_rps = 0
}

# Match Wakeup Latencies section
/Wakeup Latencies percentiles/ {
    in_wakeup = 1
    in_request = 0
    in_rps = 0
}

# Match Request Latencies section
/Request Latencies percentiles/ {
    in_wakeup = 0
    in_request = 1
    in_rps = 0
}

# Match RPS percentiles section
/RPS percentiles/ {
    in_wakeup = 0
    in_request = 0
    in_rps = 1
}

# Extract values marked with *
/\*/ {
    value = $2
    if (in_wakeup) {
        sum_wakeup += value
        count_wakeup++
    } else if (in_request) {
        sum_request += value
        count_request++
    } else if (in_rps) {
        sum_rps += value
        count_rps++
    }
}

# Output the results
END {
    print "Wakeup Latencies percentiles (usec) runtime 90 (s) (528661 total samples)"
    print "* 99.0th: " sum_wakeup / count_wakeup "         (29695 samples)"
    print ""
    print "Request Latencies percentiles (usec) runtime 90 (s) (529274 total samples)"
    print "* 99.0th: " sum_request / count_request "      (46163 samples)"
    print ""
    print "RPS percentiles (requests) runtime 90 (s) (91 total samples)"
    print "* 50.0th: " sum_rps / count_rps "       (31 samples)"
    print ""
    print "average rps: " sum_rps / count_rps
}
' schbench_output.txt
```

File: test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json
Comment (grief8) [inline test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json:14]: I think it is `requests per second`.

**Review by grief8 — CHANGES_REQUESTED** [review]
Thank you for adding the benchmarks. I have some suggestions.

**Comment by jellllly420** [comment]
hackbench running with more than one CPU fails after the rebase.

Here is the log:
<details>
  <summary>Log</summary>

```sh
~ # cd benchmark/bin/
/benchmark/bin # ./hackbench -g 8 -l 1000 -p -T
Running in threaded mode with 8 groups using 40 file descriptors each (== 320 tasks)
Each sender will pass 1000 messages of 100 bytes
[    17.561] ERROR: Uncaught panic:
        Cannot handle user page fault; Trapframe:TrapFrame {
    rax: 0x1,
    rbx: 0xffffdfffffcb5678,
    rcx: 0x4d0,
    rdx: 0x8,
    rsi: 0xffff8000049b6100,
    rdi: 0x40,
    rbp: 0xc000,
    rsp: 0xffffdfffffcb5578,
    r8: 0xffffffff884e90b0,
    r9: 0x0,
    r10: 0x0,
    r11: 0x8,
    r12: 0xd8,
    r13: 0xffffffff882bbc7f,
    r14: 0xffffdfffffcb5678,
    r15: 0xffffdfffffcb55c0,
    _pad: 0x0,
    trap_num: 0xe,
    error_code: 0x2,
    rip: 0xffffffff882cf675,
    cs: 0x30,
    rflags: 0x210086,
}.
        at /root/asterinas/ostd/src/arch/x86/trap/mod.rs:279
        on CPU 4 by thread Some(Thread { task: (Weak), data: Any { .. }, status: AtomicThreadStatus(1), priority: AtomicPriority(120), cpu_affinity: AtomicCpuSet { bits: [255] }, sched_attr: SchedAttr { policy: UnsafeCell { .. }, real_time: RealTimeAttr { prio: 99, time_slice: 35005860 }, fair: FairAttr { weight: 1024, vruntime: 0 } } })
Printing stack trace:
   1: fn 0xffffffff882c0ab0 - pc 0xffffffff882c0b3f / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb53a0;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb5210;

   2: fn 0xffffffff881bdb10 - pc 0xffffffff881bdd46 / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb53a0;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb5250;

   3: fn 0xffffffff881bdb00 - pc 0xffffffff881bdb06 / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb54d0;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb5380;

   4: fn 0xffffffff882207d0 - pc 0xffffffff882207d6 / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb54d0;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb5390;

   5: fn 0xffffffff8826ad00 - pc 0xffffffff8826ad1c / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb54d0;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb53a0;

   6: fn 0xffffffff882b27c0 - pc 0xffffffff882b2cab / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb54d0;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb53c0;

   7: fn                0x0 - pc 0xffffffff882bdfd7 / registers:

     rax 0xffffdfffffcb5088; rdx 0xffffffff8805db40; rcx                0x0; rbx 0xffffdfffffcb5678;
     rsi                0x0; rdi                0xa; rbp             0xc000; rsp 0xffffdfffffcb54d0;

[OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) ??:?
(  2) aster_nix.82604cb0e5369e36-cgu.11:?
(  3) ??:?
(  4) aster_nix_osdk_bin.a3d6357397ad4742-cgu.0:?
(  5) ??:?
(  6) ??:?
(  7) ??:?
make: *** [Makefile:200: run] Error 1

```

</details>

git bisect shows the first bad commit is 35c20620bc931f74aaf368642d42e573a99b5c25 from #1647. May need some more time to find out why.

File: test/benchmark/hackbench/summary.json
Comment (jellllly420) [reply test/benchmark/hackbench/summary.json:2]: Right. None of the results is even close. This is the current status of our scheduler. ;(

File: test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json
Comment (jellllly420) [reply test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json:8]: Thank you for the advice. 

However the schbench does give a accumulated summary at 27/28/29 lines, which means we don't have to calculate by ourselves. (And I don't think the script above is correct here; but that doesn't matter a lot.)

Actually I don't get why a `nth_occurrence` field is bad since it doesn't hurt existing configs and can save us a lot of time writing scripts. Futhermore, it seems like a common and reasonable need to pick one from the matched results when we are doing a pattern matching.

File: test/benchmark/hackbench/group8_smp16/bench_result.json
Comment (jellllly420) [reply test/benchmark/hackbench/group8_smp16/bench_result.json:13]: Thank you. It'll be fixed in next iteration.

File: test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json
Comment (jellllly420) [reply test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json:14]: Thank you. It'll be fixed in next iteration.

File: test/benchmark/README.md
Comment (tatetian) [inline test/benchmark/README.md:246]: I think a more logical ordering is to put `"search_pattern"` and `"result_index"` together and before `"nth_occurrence"`.

```json
    "result_extraction": {
        "search_pattern": "sender", // Regex or string to locate results.
        "result_index": 7 // Match index to use (e.g., 7th occurrence).
        "nth_occurrence": 1, // Which matched occurrence to use (default to 1).
    }
```

File: test/benchmark/schbench/smp1/bench_results/p99_wakeup_latency_smp1.json
Comment (tatetian) [inline test/benchmark/schbench/smp1/bench_results/p99_wakeup_latency_smp1.json:14]: What is this character?

@grief8 I am frustrated by json's inability to add comments. Having comments would be particularly useful for "result_extraction", where comments can clarify the meaning of the pattern and give examples of matched results. This "unit" field would also benefit from having a comment.

File: test/benchmark/schbench/smp1/run.sh
Comment (tatetian) [inline test/benchmark/schbench/smp1/run.sh:5]: Could you do me a favor by showing the text output of this command? I would like see the text output myself. This would be helpful to make a well-informed judgement on whether the proposed pattern matching is good.

**Review by tatetian — COMMENTED** [review]
Great work. I have some inputs for improvements.

**Comment by tatetian** [comment]
> hackbench running with more than one CPU fails after the rebase.
> 
> Here is the log:
> 
> make run RELEASE=0 ENABLE_KVM=1 SMP=8
> git bisect shows the first bad commit is [35c2062](https://github.com/asterinas/asterinas/commit/35c20620bc931f74aaf368642d42e573a99b5c25) from #1647. May need some more time to find out why.

If it takes more time to diagnose the problem, we can add the benchmarks in this PR, but not enable the ones with SMP > 1.

File: test/benchmark/schbench/smp1/bench_results/p99_wakeup_latency_smp1.json
Comment (grief8) [reply test/benchmark/schbench/smp1/bench_results/p99_wakeup_latency_smp1.json:14]: It means microseconds. It can be successfully be translated into μs on the webpage. 

It is possible to add comments but remove them before processing. For example, `"unit": "\u00b5s", # microseconds` and remove them by `sed`.

File: test/benchmark/hackbench/summary.json
Comment (grief8) [reply test/benchmark/hackbench/summary.json:2]: Thank you for the explanation.

File: test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json
Comment (grief8) [reply test/benchmark/schbench/smp1/bench_results/p50_rps_smp1.json:8]: Thanks for correcting my misunderstanding. I am OK to add `nth_occurrence` now.

File: test/benchmark/README.md
Comment (grief8) [inline test/benchmark/README.md:245]: Maybe we can emphasize it is optional to align with `runtime_config` description.
```suggestion
        "nth_occurrence": 1, // Optional. Which matched occurrence to use (default to 1).
```

File: test/benchmark/schbench/smp1/run.sh
Comment (jellllly420) [reply test/benchmark/schbench/smp1/run.sh:5]: It appends the log every 10 seconds with the latest accumulated result and prints a final summary after 90 seconds.
<details>
  <summary>raw output</summary>

```sh
/benchmark/bin # ./schbench -F 256 -n 5 -r 90
setting worker threads to 8
Wakeup Latencies percentiles (usec) runtime 10 (s) (602 total samples)
          50.0th: 68         (177 samples)
          90.0th: 1570       (236 samples)
        * 99.0th: 74880      (54 samples)
          99.9th: 375296     (6 samples)
          min=1, max=375290
Request Latencies percentiles (usec) runtime 10 (s) (596 total samples)
          50.0th: 83072      (179 samples)
          90.0th: 190720     (237 samples)
        * 99.0th: 366080     (54 samples)
          99.9th: 415232     (5 samples)
          min=16122, max=415677
RPS percentiles (requests) runtime 10 (s) (11 total samples)
          20.0th: 57         (5 samples)
        * 50.0th: 58         (1 samples)
          90.0th: 61         (5 samples)
          min=55, max=61
current rps: 55.93
Wakeup Latencies percentiles (usec) runtime 20 (s) (1209 total samples)
          50.0th: 69         (354 samples)
          90.0th: 1834       (484 samples)
        * 99.0th: 85376      (108 samples)
          99.9th: 262656     (11 samples)
          min=1, max=375290
Request Latencies percentiles (usec) runtime 20 (s) (1202 total samples)
          50.0th: 81792      (360 samples)
          90.0th: 193280     (479 samples)
        * 99.0th: 360960     (107 samples)
          99.9th: 492032     (11 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 20 (s) (21 total samples)
          20.0th: 57         (5 samples)
        * 50.0th: 59         (7 samples)
          90.0th: 61         (9 samples)
          min=55, max=61
current rps: 60.97
Wakeup Latencies percentiles (usec) runtime 30 (s) (1815 total samples)
          50.0th: 69         (530 samples)
          90.0th: 2010       (725 samples)
        * 99.0th: 85376      (163 samples)
          99.9th: 375296     (17 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 30 (s) (1811 total samples)
          50.0th: 82048      (537 samples)
          90.0th: 193792     (721 samples)
        * 99.0th: 341504     (163 samples)
          99.9th: 492032     (16 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 30 (s) (31 total samples)
          20.0th: 58         (7 samples)
        * 50.0th: 60         (21 samples)
          90.0th: 60         (0 samples)
          min=55, max=61
current rps: 56.92
Wakeup Latencies percentiles (usec) runtime 40 (s) (2430 total samples)
          50.0th: 69         (712 samples)
          90.0th: 1914       (966 samples)
        * 99.0th: 85120      (219 samples)
          99.9th: 375296     (22 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 40 (s) (2424 total samples)
          50.0th: 81536      (712 samples)
          90.0th: 194816     (974 samples)
        * 99.0th: 341504     (214 samples)
          99.9th: 482816     (22 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 40 (s) (41 total samples)
          20.0th: 59         (22 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (15 samples)
          min=55, max=61
current rps: 58.94
Wakeup Latencies percentiles (usec) runtime 50 (s) (3042 total samples)
          50.0th: 68         (879 samples)
          90.0th: 1894       (1213 samples)
        * 99.0th: 85376      (274 samples)
          99.9th: 286208     (27 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 50 (s) (3035 total samples)
          50.0th: 81536      (906 samples)
          90.0th: 194304     (1207 samples)
        * 99.0th: 346624     (273 samples)
          99.9th: 480768     (27 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 50 (s) (51 total samples)
          20.0th: 59         (28 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 61         (23 samples)
          min=55, max=61
current rps: 60.98
Wakeup Latencies percentiles (usec) runtime 60 (s) (3650 total samples)
          50.0th: 68         (1070 samples)
          90.0th: 1874       (1450 samples)
        * 99.0th: 89216      (329 samples)
          99.9th: 286208     (33 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 60 (s) (3642 total samples)
          50.0th: 81536      (1101 samples)
          90.0th: 193792     (1442 samples)
        * 99.0th: 342528     (326 samples)
          99.9th: 480768     (33 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 60 (s) (61 total samples)
          20.0th: 59         (33 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (22 samples)
          min=55, max=61
current rps: 60.94
Wakeup Latencies percentiles (usec) runtime 70 (s) (4265 total samples)
          50.0th: 68         (1293 samples)
          90.0th: 1846       (1687 samples)
        * 99.0th: 89216      (384 samples)
          99.9th: 262656     (38 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 70 (s) (4259 total samples)
          50.0th: 81536      (1285 samples)
          90.0th: 193792     (1687 samples)
        * 99.0th: 348672     (382 samples)
          99.9th: 480768     (38 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 70 (s) (71 total samples)
          20.0th: 59         (37 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (27 samples)
          min=55, max=61
current rps: 60.97
Wakeup Latencies percentiles (usec) runtime 80 (s) (4872 total samples)
          50.0th: 68         (1437 samples)
          90.0th: 1922       (1927 samples)
        * 99.0th: 96896      (439 samples)
          99.9th: 349696     (44 samples)
          min=1, max=502160
Request Latencies percentiles (usec) runtime 80 (s) (4865 total samples)
          50.0th: 81536      (1468 samples)
          90.0th: 193792     (1938 samples)
        * 99.0th: 346624     (430 samples)
          99.9th: 484864     (44 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 80 (s) (81 total samples)
          20.0th: 59         (41 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (33 samples)
          min=55, max=61
current rps: 60.92
Wakeup Latencies percentiles (usec) runtime 90 (s) (5481 total samples)
          50.0th: 68         (1611 samples)
          90.0th: 1894       (2190 samples)
        * 99.0th: 89216      (494 samples)
          99.9th: 290304     (49 samples)
          min=1, max=502160
Request Latencies percentiles (usec) runtime 90 (s) (5474 total samples)
          50.0th: 81536      (1655 samples)
          90.0th: 193280     (2169 samples)
        * 99.0th: 348672     (489 samples)
          99.9th: 489984     (49 samples)
          min=16122, max=606819
RPS percentiles (requests) runtime 90 (s) (91 total samples)
          20.0th: 59         (47 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (36 samples)
          min=55, max=61
current rps: 59.94
Wakeup Latencies percentiles (usec) runtime 90 (s) (5482 total samples)
          50.0th: 68         (1611 samples)
          90.0th: 1906       (2191 samples)
        * 99.0th: 90496      (494 samples)
          99.9th: 290304     (49 samples)
          min=1, max=502160
Request Latencies percentiles (usec) runtime 90 (s) (5482 total samples)
          50.0th: 81536      (1658 samples)
          90.0th: 193280     (2173 samples)
        * 99.0th: 348672     (490 samples)
          99.9th: 489984     (49 samples)
          min=16122, max=606819
RPS percentiles (requests) runtime 90 (s) (91 total samples)
          20.0th: 59         (47 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (36 samples)
          min=55, max=61
average rps: 60.91
/benchmark/bin # 
```

</details>

Comment (tatetian) [reply test/benchmark/schbench/smp1/run.sh:5]: @jellllly420 Thanks for providing the sample output.

Now I understood why you would want to match the 28-th occurrences with the following configuration.

```json
    "result_extraction": {
        "search_pattern": "\\*",
        "nth_occurrence": 28,
    }
```

But having this magic number of 28 as part of the pattern matching rules is not good for _maintainability_. First, it is _unreadable_. Without seeing a sample output, people will have no idea why it is the 28-th occurence that is the result we want. Second, it is _fragile_. We could easily change our mind in the future to run for more time by adjusting the `-r` parameter. Every time we do so, we need to remember to adjust the `nth_occurrence`. This is exactly the reason why #1725 replaces all hardcoded IP addresses with regular expressions.

But if we don't introduce `nth_occurrence`, how we can match the result patterns that may appear in the output text for _multiple times_? Luckily, in this specific case of `schbench`, we don't have to.

```
-r (--runtime): How long to run before exiting (seconds, def: 30)
-i (--intervaltime): interval for printing latencies (seconds, def: 10) Print a report about our latencies and RPS ever N seconds.
```

We should change the command to specify the same time for both parameters. This means that the accumulated result will only show once.

```bash
RUNTIME=90
/benchmark/bin/schbench -r $RUNTIME -i RUNTIME -F 256 -n 5
```

File: test/benchmark/schbench/smp1/bench_results/p99_wakeup_latency_smp1.json
Comment (grief8) [reply test/benchmark/schbench/smp1/bench_results/p99_wakeup_latency_smp1.json:14]: Yes, it's time to adopt a new format. YAML is an excellent choice for its readability and flexibility. We can leverage `yq` to decode and manipulate YAML files, as it is designed specifically for YAML and operates in a manner similar to `jq`, making it a seamless transition for those familiar with JSON processing. It will be discussed in another PR.

File: test/benchmark/schbench/smp1/run.sh
Comment (jellllly420) [reply test/benchmark/schbench/smp1/run.sh]: Thank you for the advice. Now I changed the run script like below
```sh
RUNTIME=90
REPORT_INTERVAL=$((RUNTIME+10))

/benchmark/bin/schbench -F 256 -n 5 -r $RUNTIME -i $REPORT_INTERVAL
```
Then schbench will only print a final summary after `$RUNTIME` seconds, which looks like below
```sh
Wakeup Latencies percentiles (usec) runtime 90 (s) (255474 total samples)
          50.0th: 7          (66640 samples)
          90.0th: 20         (68973 samples)
        * 99.0th: 272        (21770 samples)
          99.9th: 1266       (2301 samples)
          min=1, max=10481
Request Latencies percentiles (usec) runtime 90 (s) (255755 total samples)
          50.0th: 17120      (85972 samples)
          90.0th: 17568      (65273 samples)
        * 99.0th: 19616      (22817 samples)
          99.9th: 26464      (2249 samples)
          min=10794, max=48606
RPS percentiles (requests) runtime 90 (s) (9 total samples)
          20.0th: 2996       (3 samples)
        * 50.0th: 3004       (2 samples)
          90.0th: 3020       (4 samples)
          min=2982, max=3021
average rps: 3009.10
```
But still we cannot distinguish between the p99 wakeup latency and the p99 request latency with the current matching mechanism. We can (1) keep the `nth_occurrence` field (now it's stable, just 1/2/3, and won't be affected by any parameter change) or (2) reconstruct the result in run script like @grief8 did in this [comment](https://github.com/asterinas/asterinas/pull/1677#discussion_r1890120398)

The latest code adopted the first approach but I'm totally OK with switching to the second.

Comment (tatetian) [reply test/benchmark/schbench/smp1/run.sh]: I am fine with `nth_occurrence`, too.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**
